app/src-docs/config.md
<!-- markdownlint-disable -->

<a href="../src/github_runner_image_builder/config.py#L0"><img align="right" style="float:right;" src="https://img.shields.io/badge/-source-cccccc?style=flat-square"></a>

# <kbd>module</kbd> `config`
Module containing configurations. 

**Global Variables**
---------------
- **ARCHITECTURES_ARM64**
- **ARCHITECTURES_X86**
- **LTS_IMAGE_VERSION_TAG_MAP**
- **BASE_CHOICES**
- **IMAGE_DEFAULT_APT_PACKAGES**
- **LOG_LEVELS**

---

<a href="../src/github_runner_image_builder/config.py#L47"><img align="right" style="float:right;" src="https://img.shields.io/badge/-source-cccccc?style=flat-square"></a>

## <kbd>function</kbd> `get_supported_arch`

```python
get_supported_arch() → <enum 'Arch'>
```

Get current machine architecture. 



**Raises:**
 
 - <b>`UnsupportedArchitectureError`</b>:  if the current architecture is unsupported. 



**Returns:**
 
 - <b>`Arch`</b>:  Current machine architecture. 


---

<a href="../src/github_runner_image_builder/config.py#L18"><img align="right" style="float:right;" src="https://img.shields.io/badge/-source-cccccc?style=flat-square"></a>

## <kbd>class</kbd> `Arch`
Supported system architectures. 



**Attributes:**
 
 - <b>`ARM64`</b>:  Represents an ARM64 system architecture. 
 - <b>`X64`</b>:  Represents an X64/AMD64 system architecture. 





---

<a href="../src/github_runner_image_builder/config.py#L66"><img align="right" style="float:right;" src="https://img.shields.io/badge/-source-cccccc?style=flat-square"></a>

## <kbd>class</kbd> `BaseImage`
The ubuntu OS base image to build and deploy runners on. 



**Attributes:**
 
 - <b>`JAMMY`</b>:  The jammy ubuntu LTS image. 
 - <b>`NOBLE`</b>:  The noble ubuntu LTS image. 





---

<a href="../src/github_runner_image_builder/config.py#L141"><img align="right" style="float:right;" src="https://img.shields.io/badge/-source-cccccc?style=flat-square"></a>

## <kbd>class</kbd> `ScriptConfig`
The custom setup script configurations. 



**Attributes:**
 
 - <b>`script_url`</b>:  The external setup bash script URL. 
 - <b>`script_secrets`</b>:  The space separated external secrets to load before running external             script_url. e.g. "SECRET_ONE=HELLO SECRET_TWO=WORLD" 

<a href="../<string>"><img align="right" style="float:right;" src="https://img.shields.io/badge/-source-cccccc?style=flat-square"></a>

### <kbd>method</kbd> `__init__`

```python
__init__(script_url: ParseResult | None, script_secrets: dict[str, str]) → None
```









---

<a href="../src/github_runner_image_builder/config.py#L155"><img align="right" style="float:right;" src="https://img.shields.io/badge/-source-cccccc?style=flat-square"></a>

## <kbd>class</kbd> `ImageConfig`
The build image configuration values. 



**Attributes:**
 
 - <b>`arch`</b>:  The architecture of the target image. 
 - <b>`base`</b>:  The ubuntu base OS of the image. 
 - <b>`microk8s`</b>:  The MicroK8s snap channel to install. 
 - <b>`juju`</b>:  The Juju channel to install and bootstrap. 
 - <b>`runner_version`</b>:  The GitHub runner version to install on the VM. Defaults to latest. 
 - <b>`script_config`</b>:  The custom setup script configurations. 
 - <b>`name`</b>:  The image name to upload on OpenStack. 

<a href="../<string>"><img align="right" style="float:right;" src="https://img.shields.io/badge/-source-cccccc?style=flat-square"></a>

### <kbd>method</kbd> `__init__`

```python
__init__(
    arch: Arch,
    base: BaseImage,
    microk8s: str,
    juju: str,
    runner_version: str,
    script_config: ScriptConfig,
    name: str
) → None
```











docs/how-to/configure-base-image.md
# How to configure `base-image`

By default, github-runner-image-builder uses the `noble` Ubuntu OS base. To choose a different
base, you can use the `base-image` configuration option.

```
BASE_IMAGE=jammy
OPENSTACK_AUTH_URL=<openstack-auth-url, e.g. http://my-openstack-deployment/openstack-keystone>
OPENSTACK_PASSWORD=<openstack project password>
OPENSTACK_PROJECT_DOMAIN_NAME=<openstack project domain name>
OPENSTACK_PROJECT_NAME=<openstack project name>
OPENSTACK_USER_DOMAIN_NAME=<openstack user domain name>
OPENSTACK_USERNAME=<openstack username>

juju deploy github-runner-image-builder \
--config base-image=$BASE_IMAGE
--config openstack-auth-url=$OPENSTACK_AUTH_URL \
--config openstack-password=$OPENSTACK_PASSWORD \
--config openstack-project-domain-name=$OPENSTACK_PROJECT_DOMAIN_NAME \
--config openstack-project-name=$OPENSTACK_PROJECT_NAME \
--config openstack-user-domain-name=$OPENSTACK_USER_DOMAIN_NAME \
--config openstack-user-name=$OPENSTACK_USERNAME
```

Currently, jammy (22.04) and noble (24.04) are supported.


RELEASE-NOTES.md
## [#27 feat: external cloud](https://github.com/canonical/github-runner-image-builder-operator/pull/27) (2021-09-13)

> Builds GitHub runner images on OpenStack VMs.

### Upgrade Steps
* Requires enough capacity on the OpenStack project cloud to launch builder VMs.
* Requires GitHub Runner revision 249 and above.

### Breaking Changes
* None

### New Features
* None

### Bug Fixes
* None

### Performance Improvements
* None

### Other Changes
* None


app/src/github_runner_image_builder/config.py
# Copyright 2025 Canonical Ltd.
# See LICENSE file for licensing details.

"""Module containing configurations."""

import dataclasses
import itertools
import logging
import platform
import urllib.parse
from enum import Enum
from pathlib import Path
from typing import Literal

from github_runner_image_builder.errors import UnsupportedArchitectureError


class Arch(str, Enum):
    """Supported system architectures.

    Attributes:
        ARM64: Represents an ARM64 system architecture.
        X64: Represents an X64/AMD64 system architecture.
    """

    ARM64 = "arm64"
    X64 = "x64"

    def to_openstack(self) -> str:
        """Convert the architecture to OpenStack compatible arch string.

        Returns:
            The architecture string.
        """  # noqa: DCO050 the ValueError is an unreachable code.
        match self:
            case Arch.ARM64:
                return "aarch64"
            case Arch.X64:
                return "x86_64"
        raise ValueError  # pragma: nocover


ARCHITECTURES_ARM64 = {"aarch64", "arm64"}
ARCHITECTURES_X86 = {"x86_64"}


def get_supported_arch() -> Arch:
    """Get current machine architecture.

    Raises:
        UnsupportedArchitectureError: if the current architecture is unsupported.

    Returns:
        Arch: Current machine architecture.
    """
    arch = platform.machine()
    match arch:
        case arch if arch in ARCHITECTURES_ARM64:
            return Arch.ARM64
        case arch if arch in ARCHITECTURES_X86:
            return Arch.X64
        case _:
            raise UnsupportedArchitectureError()


class BaseImage(str, Enum):
    """The ubuntu OS base image to build and deploy runners on.

    Attributes:
        JAMMY: The jammy ubuntu LTS image.
        NOBLE: The noble ubuntu LTS image.
    """

    JAMMY = "jammy"
    NOBLE = "noble"

    @classmethod
    def get_version(cls, base: "BaseImage") -> Literal["22.04", "24.04"]:
        """Change the codename to version tag.

        Args:
            base: The base image to get the version number for.

        Return:
            The release version of the current base image.
        """
        match base:
            case BaseImage.JAMMY:
                return "22.04"
            case BaseImage.NOBLE:
                return "24.04"

    @classmethod
    def from_str(cls, tag_or_name: str) -> "BaseImage":
        """Retrieve the base image tag from input.

        Args:
            tag_or_name: The base image string option.

        Returns:
            The base image configuration of the app.
        """
        if tag_or_name in LTS_IMAGE_VERSION_TAG_MAP:
            return cls(LTS_IMAGE_VERSION_TAG_MAP[tag_or_name])
        return cls(tag_or_name)


LTS_IMAGE_VERSION_TAG_MAP = {"22.04": BaseImage.JAMMY.value, "24.04": BaseImage.NOBLE.value}
BASE_CHOICES = tuple(
    itertools.chain.from_iterable((tag, name) for (tag, name) in LTS_IMAGE_VERSION_TAG_MAP.items())
)
IMAGE_OUTPUT_PATH = Path("compressed.img")

IMAGE_DEFAULT_APT_PACKAGES = [
    "build-essential",
    "docker.io",
    "gh",
    "jq",
    "npm",
    "python3-dev",
    "python3-pip",
    "python-is-python3",
    "shellcheck",
    "tar",
    "time",
    "unzip",
    "wget",
]

_LOG_LEVELS = (logging.DEBUG, logging.INFO, logging.WARNING, logging.ERROR)
LOG_LEVELS = tuple(
    str(level)
    for level in itertools.chain(
        _LOG_LEVELS,
        (logging.getLevelName(level) for level in _LOG_LEVELS),
        (logging.getLevelName(level).lower() for level in _LOG_LEVELS),
    )
)


@dataclasses.dataclass
class ScriptConfig:
    """The custom setup script configurations.

    Attributes:
        script_url: The external setup bash script URL.
        script_secrets: The space separated external secrets to load before running external \
            script_url. e.g. "SECRET_ONE=HELLO SECRET_TWO=WORLD"
    """

    script_url: urllib.parse.ParseResult | None
    script_secrets: dict[str, str]


@dataclasses.dataclass
class ImageConfig:
    """The build image configuration values.

    Attributes:
        arch: The architecture of the target image.
        base: The ubuntu base OS of the image.
        microk8s: The MicroK8s snap channel to install.
        juju: The Juju channel to install and bootstrap.
        runner_version: The GitHub runner version to install on the VM. Defaults to latest.
        script_config: The custom setup script configurations.
        name: The image name to upload on OpenStack.
    """

    arch: Arch
    base: BaseImage
    microk8s: str
    juju: str
    runner_version: str
    script_config: ScriptConfig
    name: str
